src/hotspot/os/posix/malloctrace/mallocTrace2.cpp

```diff
@@ -0,0 +1,2675 @@
+  }
+
+  for (int i = 0; i < NR_OF_STACK_MAPS; ++i) {
+    if (pthread_mutex_init(&_stack_maps_lock[i]._lock, NULL) != 0) {
+      fatal("Could not initialize lock");
+    }
+  }
+
+  for (int i = 0; i < NR_OF_ALLOC_MAPS; ++i) {
+    if (pthread_mutex_init(&_alloc_maps_lock[i]._lock, NULL) != 0) {
+      fatal("Could not initialize lock");
+    }
+  }
+}
+
+bool MallocStatisticImpl::enable(outputStream* st, TraceSpec const& spec) {
+  initialize();
+  Locker lock(_malloc_stat_lock);
+
+  if (_enabled) {
+    if (spec._force) {
+      _enabled = false;
+      setup_hooks(NULL, st);
+      cleanup();
+
+      st->print_raw_cr("Disabling already running trace first.");
+    } else {
+      st->print_raw_cr("Malloc statistic is already enabled!");
+
+      return false;
+    }
+  }
+
+  if (_shutdown) {
+    st->print_raw_cr("Malloc statistic is already shut down!");
+
+    return false;
+  }
+
+  if (spec._stack_depth < 2 || spec._stack_depth > MAX_FRAMES) {
+    st->print_cr("The given stack depth %d is outside of the valid range [%d, %d]",
+                 spec._stack_depth, 2, MAX_FRAMES);
+
+    return false;
+  }
+
+  // Get the backtrace function if needed.
+  if (spec._use_backtrace && !_tried_to_load_backtrace) {
+    _tried_to_load_backtrace = true;
+
+#if defined(__APPLE__)
+    // Try libunwind first on mac.
+    _backtrace = (backtrace_func_t*) dlsym(RTLD_DEFAULT, "unw_backtrace");
+    _backtrace_name = "backtrace (libunwind)";
+
+    if (_backtrace == NULL) {
+      _backtrace = (backtrace_func_t*) dlsym(RTLD_DEFAULT, "backtrace");
+      _backtrace_name = "backtrace";
+    }
+#else
+    _backtrace = (backtrace_func_t*) dlsym(RTLD_DEFAULT, "backtrace");
+    _backtrace_name = "backtrace";
+
+    if (_backtrace == NULL) {
+      // Try if we have libunwind installed.
+      char ebuf[512];
+      void* libunwind = os::dll_load(MallocTraceUnwindLibName, ebuf, sizeof ebuf);
+
+      if (libunwind != NULL) {
+        _backtrace = (backtrace_func_t*) dlsym(libunwind, "unw_backtrace");
+        _backtrace_name = "backtrace (libunwind)";
+      }
+    }
+#endif
+
+    // Clear dlerror.
+    dlerror();
+
+    if (_backtrace != NULL) {
+      // Trigger initialization needed.
+      void* tmp[1];
+      _backtrace(tmp, 1);
+    }
+  }
+
+  _track_free = spec._track_free;
+  _detailed_stats = spec._detailed_stats;
+
+  if (_track_free) {
+    st->print_raw_cr("Tracking live memory.");
+  } else {
+    st->print_raw_cr("Tracking all allocated memory.");
+  }
+
+  if (_detailed_stats) {
+    st->print_raw_cr("Collecting detailed statistics.");
+  }
+
+  int only_nth = MIN2(1000, MAX2(1, spec._only_nth));
+
+  if (only_nth > 1) {
+    for (int i = 1; i < 63; ++i) {
+      uint64_t pow = ((uint64_t) 1) << i;
+      _to_track_limit = pow / only_nth;
+      double diff = pow / (double) _to_track_limit - only_nth;
+
+      _to_track_mask = pow - 1;
+
+      if (diff > -0.1 && diff < 0.1) {
+        break;
+      }
```

Comment:
That's what I've done initially. But this version allows to approximate non-power-of-2 values too and the check itself is nearly as fast (masking and comparing with threshold).

Reply:
Yes, the loop is really not needed (expect if smaller masks are faster to use ;)). Why 42 and not 62?

Reply:
42 (actually 43) is sufficient to even trace every 1,000,000th malloc with an error < 0.1 (and it is such a special number too).
There's a dependency to the maximal hash value. You could introduce a  constant for it and derive `pow` from it.
62, 42, maximal hash, what ever you like better. It would be nice though to get rid of the loop.